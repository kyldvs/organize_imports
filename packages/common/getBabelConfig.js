/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

const DEFAULT_CONFIG = {
  sourceType: 'module',
  plugins: [
    'syntax-async-functions',
    'syntax-flow',
    'syntax-jsx',
    'syntax-object-rest-spread',
    'syntax-trailing-function-commas',
  ],
};

export default function getBabelConfig(config) {
  if (!config) {
    return DEFAULT_CONFIG;
  }
  const result = {};
  const allKeys = [].concat(Object.keys(config), Object.keys(DEFAULT_CONFIG));
  allKeys.forEach(key => {
    if (config[key]) {
      if (Array.isArray(DEFAULT_CONFIG[key])) {
        result[key] = [].concat(DEFAULT_CONFIG[key], config[key]);
      } else {
        result[key] = config[key];
      }
    } else {
      result[key] = DEFAULT_CONFIG[key];
    }
  });
  return result;
};
