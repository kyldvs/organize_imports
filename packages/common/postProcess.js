/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

export default function postProcess(code) {
  return code.replace(/\s*\$\$NEWLINE\$\$;\s*/g, '\n');
};
