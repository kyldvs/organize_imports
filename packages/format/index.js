/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

export default function format(options) {
  return function babel_plugin({types: t}) {
    return {
      visitor: {
      },
    };
  };
}
