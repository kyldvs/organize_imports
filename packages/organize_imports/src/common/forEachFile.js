/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import fs from 'fs';
import path from 'path';

export default function forEachFile(dir, fn) {
  getFiles(dir, dir).forEach(x => fn(x));
}

function getFiles(firstFilePath, originalFilePath) {
  let arr = [];
  const filePaths = fs.readdirSync(originalFilePath);
  filePaths.forEach(fileName => {
    const filePath = path.join(originalFilePath, fileName);
    const stats = fs.statSync(filePath);
    if (stats.isDirectory()) {
      arr = [].concat(arr, getFiles(firstFilePath, filePath));
    } else if (stats.isFile()) {
      arr.push({
        name: fileName,
        relativeName: path.relative(firstFilePath, filePath),
        contents: fs.readFileSync(filePath, 'utf8'),
      });
    }
  });
  return arr;
}
