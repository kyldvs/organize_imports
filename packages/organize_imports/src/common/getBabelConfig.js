/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

const DEFAULT_CONFIG = {
  sourceType: 'module',
  plugins: [
    require('babel-plugin-syntax-async-functions'),
    require('babel-plugin-syntax-flow'),
    require('babel-plugin-syntax-jsx'),
    require('babel-plugin-syntax-object-rest-spread'),
    require('babel-plugin-syntax-trailing-function-commas'),
  ],
};

export default function getBabelConfig(config) {
  if (!config) {
    return DEFAULT_CONFIG;
  }
  const result = {};
  const allKeys = [].concat(Object.keys(config), Object.keys(DEFAULT_CONFIG));
  allKeys.forEach(key => {
    if (config[key]) {
      if (Array.isArray(DEFAULT_CONFIG[key])) {
        result[key] = [].concat(DEFAULT_CONFIG[key], config[key]);
      } else {
        result[key] = config[key];
      }
    } else {
      result[key] = DEFAULT_CONFIG[key];
    }
  });
  return result;
};
