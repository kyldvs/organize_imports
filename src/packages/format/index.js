/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import RequireGroups from '../common/RequireGroups';

import getRequireIDs from '../common/getRequireIDs';
import isRequire from '../common/isRequire';
import match from '../common/match';
import reprintRequireGroups from '../common/reprintRequireGroups';

export default function format(options) {
  return function babel_plugin({types: t}) {
    return {
      visitor: {
        Program(path) {

          const pre = [];
          const requires = [];
          const post = [];

          // Partition all the body nodes, into pre-require-block,
          // require-block, and post-require-block.
          path.node.body.forEach(node => {
            if (isRequire(node)) {
              requires.push(node);
            } else {
              if (requires.length > 0) {
                post.push(node);
              } else {
                pre.push(node);
              }
            }
          });

          // Transform require nodes as appropriate.
          requires.forEach(node => {
            // TODO: Sort specifiers and destructures.
          });

          // Sort all the requires, this will prevent us needing to re-order
          // the groups later.
          requires.sort((n1, n2) => {
            const ids1 = getRequireIDs(n1);
            const ids2 = getRequireIDs(n2);

            // This should never happen.
            if (ids1.length === 0 || ids2.length === 0) {
              return 0;
            }

            for (let i = 0; i < ids1.length && i < ids2.length; i++) {
              if (ids1[i] < ids2[i]) {
                return -1;
              } else if (ids1[i] > ids2[i]) {
                return 1;
              }
            }

            // Since they are all equal to this point we can compare based on
            // the number of ids.
            if (ids1.length < ids2.length) {
              return -1;
            } else if (ids1.length > ids2.length) {
              return 1;
            } else {
              return 0;
            }
          });

          // Create an empty array for each group.
          const groups = RequireGroups.map(() => []);
          const unknowns = [];

          // For each require add it to the correct group.
          requires.forEach(node => {
            for (let i = 0; i < RequireGroups.length; i++) {
              if (RequireGroups[i].some(pattern => match(node, pattern))) {
                groups[i].push(node);
                return;
              }
            }

            // If we make it to here that's bad, but let's not drop the node.
            unknowns.push(node);
          });

          const newRequires = reprintRequireGroups(groups);

          // Sort each group.
          path.node.body = [... pre, ...newRequires, ...post];
        },
      },
    };
  };
}
