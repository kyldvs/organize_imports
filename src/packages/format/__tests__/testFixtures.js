/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import format from '../index';
import * as babel from 'babel-core';
import forEachFile from '../../common/forEachFile';
import getBabelConfig from '../../common/getBabelConfig';
import path from 'path';
import postProcess from '../../common/postProcess';

const FIXTURES_DIR = path.join(__dirname, '..', '__fixtures__');

forEachFile(FIXTURES_DIR, ({relativeName, contents}) => {
  test(relativeName, () => {
    expect(run(contents)).toMatchSnapshot();
  });
});

function run(code, options) {
  return postProcess(babel.transform(code, getBabelConfig({
    plugins: [
      format(options),
    ],
  })).code);
}
