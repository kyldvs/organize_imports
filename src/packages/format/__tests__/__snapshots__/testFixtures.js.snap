exports[`test simple/test1.js 1`] = `
"import type D from \'D\';
import type e from \'e\';

const B = require(\'B\');
const G = require(\'G\');

const a = require(\'a\');
const c = require(\'c\');

const { F } = require(\'F\');

a;
B;
c;"
`;
